### ISSUE TITLE
`toJSON` keyword?

### ISSUE THREAD

**Issue by Gabriella439** [dhall-lang/dhall-lang#336]
I'm opening up an issue to discuss this comment:

https://github.com/dhall-lang/dhall-lang/issues/125#issuecomment-450601520

... which is also related to this repository:

https://github.com/akshaymankar/dhall-recursive-json

The informal proposal is to add a new `JSON` primitive type and a `toJSON` keyword that converts any Dhall expression to the corresponding JSON expression.  The repository linked above has the `toJSON` keyword taking a type argument followed by the expression to convert to JSON, but we can simplify that a bit to remove the type argument since we can infer the type from the expression to convert (similar to how the `Some` keyword behaves).

Conceptually the `JSON` type is an opaque type (i.e. similar to how we treat `Double`).  You can create it using `toJSON` but you can't do anything with it within the language.  The only interesting thing you can do with it is embed it within a larger expression that you convert to JSON.  In other words, this is legal:

```haskell
let x : JSON = toJSON 1
let y : JSON = toJSON True
in  toJSON [ x, y ]
```

... which would correspond to the following JSON expression:

```json
[ 1, true ]
```

Note that normalizing `toJSON` just normalizes its argument.  In other words, if we were to normalize the above Dhall expression the normal form would be:

```haskell
toJSON [ toJSON 1, toJSON True ]
```

Type-checking `toJSON` would essentially entail verifying that the Dhall expression is convertable to JSON (which can be made precise, but this is just an informal exposition of the idea).  If that check succeeds then the inferred type is `JSON`.

The non-trivial aspect to this is that each language binding would now be responsible for converting any value of type `JSON` to a JSON expression within the host language.  For example, the Haskell bindings to Dhall would now have an `Interpret` instance for decoding into the Haskell representation of a `JSON` value:

```haskell
instance Interpret Data.Aeson.Value where ...
```

I'll respond to this in a separate comment with my own thoughts on this proposal.

cc: @akshaymankar

**Comment by Gabriel439** [comment]
My view on this is that I'm in favor but currently I'm trying to pause changes to the language evolution to give the `dhall-clj`/`dhall-purescript`/`dhall-python` implementations a chance to catch up with the standard.

The main reasons I'm in favor are:

* `dhall` is positioning itself as a YAML replacement and in order to do so it should natively support JSON.  This also dovetails with the proposal to support importing values `as JSON`: https://github.com/dhall-lang/dhall-lang/issues/121
* It's very simple to implement

The main issue I can foresee is that an implementation might have to pick a standard JSON type to decode into.  For some language ecosystems this is easy because there is one right way to do JSON, such as Python (which has a standard JSON library) or Haskell (which has a de-factor standard JSON library in `aeson`).  In other ecosystems this is not the case, such as Scala (which has multiple competing JSON libraries).

However, even in the worst case scenario like Scala an implementation can either (A) define its own a bare-bones JSON data structure to decode into or (B) just be opinionated and pick a JSON library to use.

Note that we have turned down some similar proposals such as:

* Native support for IP addresses: https://github.com/dhall-lang/dhall-lang/issues/217
* Native support for timestamps: https://github.com/dhall-lang/dhall-lang/issues/80

My main reasons for supporting JSON compared to the other primitive types is that:

* Supporting JSON natively is necessary to properly displace YAML

  YAML does not support IP or timestamp primitive types either, so we lose nothing comparatively if we don't support them.  However, if we ever decide to also displace TOML then at some point we will need to support timestamps natively, too.

* We get a lot of "bang for our buck" standardizing support for JSON compared to other primitive types

  In particular, if we combine this proposal with support for importing expressions `as JSON` then Dhall effectively can be used as an all-in-one solution for JSON data muxing.  Also, Dhall then provides a very nice migration story for existing JSON pipelines.

However, like I mentioned above, I'd like to wait until we have more native language bindings to Dhall before trying this since I think native language support will have a bigger impact on Dhall adoption than JSON support.

**Comment by jneira** [comment]
I am not sure about this: wouldn't be enough to use the existent utility `dhall-json` (with the changes proposed in https://github.com/dhall-lang/dhall-lang/issues/125#issuecomment-376925304)? 

This could lead to users to ask for other formats, like `toYAML` itself (the main format to replace with dhall, right?) and extend the support to import `as JSON` that could be lead to standarize an opitionated way to make conversions.

IMHO conversions with other formats (any of them, in the present or future) should be keeping in official tooling, to try keep dhall itself as simple as possible.

**Comment by Gabriel439** [comment]
Thinking about this more, there is a much simpler way to do this.  You can (almost) implement the Dhall to JSON conversion in pure Dhall

I have a branch illustrating how to do this here:

https://github.com/dhall-lang/dhall-lang/tree/gabriel/json_pure_dhall

... and here is an example program to assemble and render a JSON expression:

```haskell
let JSON =
      https://raw.githubusercontent.com/dhall-lang/dhall-lang/gabriel/json_pure_dhall/Prelude/JSON/package.dhall sha256:76a75f75e9ab6cd78d18b123f10946cc7778a3d734e50978d635505cae8c062a

let value =
      JSON.object
      [ { mapKey =
            "foo"
        , mapValue =
            JSON.array [ JSON.bool True, JSON.number 1.0 ]
        }
      , { mapKey = "bar", mapValue = JSON.string "!" }
      ]

in  JSON.render value
```

```bash
$ dhall-to-text <<< './example.dhall' 
{ "foo": [ true, 1.0 ], "bar": "!" }
```

This gives us most of the benefits of the proposal without any changes to the language.  With a little extra work we could even render pretty-printed JSON, too, entirely in Dhall.  This would effectively obsolete the `dhall-to-json` executable.

The only missing piece is that Dhall needs a new `Text/show` primitive to escape its own `Text` literals so that we can render Dhall  strings as JSON strings.  If we have that then we could implement Dhall conversions to all sorts of configuration formats (including YAML) without having to extend the language.  Also, `Text/show` is a useful general-purpose primitive to add to the language, consistent with the other `/show` primitives.

Or in other words, this is essentially reusing the same trick as the `dhall-nethack` project to embed support for JSON within the language by implementing the abstract syntax tree and the rendering function in pure Dhall.

**Comment by ocharles** [comment]
I would not call it show as that is a very Haskell specific term. How about
'Text/escape'?

On 4 Jan 2019 5:00 pm, "Gabriel Gonzalez" <notifications@github.com> wrote:

Thinking about this more, there is a much simpler way to do this. You can
(almost) implement the Dhall to JSON conversion in pure Dhall

I have a branch illustrating how to do this here:

https://github.com/dhall-lang/dhall-lang/tree/gabriel/json_pure_dhall

... and here is an example program to assemble and render a JSON expression:

let JSON =
      https://raw.githubusercontent.com/dhall-lang/dhall-lang/gabriel/json_pure_dhall/Prelude/JSON/package.dhall
sha256:76a75f75e9ab6cd78d18b123f10946cc7778a3d734e50978d635505cae8c062a
let value =
      JSON.object
      [ { mapKey =
            "foo"
        , mapValue =
            JSON.array [ JSON.bool True, JSON.number 1.0 ]
        }
      , { mapKey = "bar", mapValue = JSON.string "!" }
      ]
in  JSON.render value

$ dhall-to-text <<< './example.dhall'
{ "foo": [ true, 1.0 ], "bar": "!" }

This gives us most of the benefits of the proposal without any changes to
the language. With a little extra work we could even render pretty-printed
JSON, too, entirely in Dhall. This would effectively obsolete the
dhall-to-json executable.

The only missing piece is that Dhall needs a new Text/show primitive to
escape its own Text literals so that we can render Dhall strings as JSON
strings. If we have that then we could implement Dhall conversions to all
sorts of configuration formats (including YAML) without having to extend
the language. Also, Text/show is a useful general-purpose primitive to add
to the language, consistent with the other /show primitives.

Or in other words, this is essentially reusing the same trick as the
dhall-nethack project to embed support for JSON within the language by
implementing the abstract syntax tree and the rendering function in pure
Dhall.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/dhall-lang/dhall-lang/issues/336#issuecomment-451503204>,
or mute the thread

**Comment by Gabriel439** [comment]
@ocharles: There are two reasons I prefer to name this function `Text/show`:

The first reason is that it's consistent with the naming of the `Natural/show`, `Integer/show`, and `Double/show` utilities, which render the equivalent Dhall representation.

The second reason is that in my mind there is a slight semantic difference between `Text/escape` and `Text/show`, which is that I would expect the former to render the escaped text without surrounding quotes and the latter to render with surrounding quotes.  In other words:

```haskell
Text/escape "abc\ndef" = "abc\\ndef"

Text/show "abc\ndef" = "\"abc\\ndef\""
```

What I was proposing to add to the language was the latter functionality (with surrounding quotes, so that the rendered result is both valid Dhall source code and also a valid JSON string literal).

The main issue I foresee with this is that JSON does not support the `\$` escape sequence that Dhall does, so `Text/show` would have to replace the dollar sign in `"\${""}"` with `\u0024` if it precedes a `{` in order to be both JSON-compatible and Dhall-compatible.  Or in other words:

```haskell
Text/show "\${""}" = "\"\\u0024{\"\"}\""
```

**Comment by akshaymankar** [comment]
I tried using the `json_pure_dhall` branch and realized converting arbitrary records to JSON is not easy. Users have to write conversion for each type of record to `List {mapKey: Text, mapValue: JSON}` which is very tedious, especially if records are big.

Also, I think rendering YAML as text might get very tricky in dhall. Although, some other tool could be used to convert JSON to YAML.

On the other hand, my original hack of putting `toJSON` and `JSON` in `startingContext` doesn't work as well as I thought it would because I cannot use them directly in imported files. I think this is because starting context is only applied to inital file. I can, however, wrap all files which render JSONs with `λ(JSON : Type) → λ(toJSON : (Value: Type) → (v: Value) → JSON)`. Its not very nice, but I can deal with it until a better solution is available.

So, I think it would be best (at least for generating JSONs and YAMLs) `toJSON` was recognized by dhall as a builtin function and `JSON` was recognized as builtin Type.

**Comment by Gabriel439** [comment]
@akshaymankar: You can customize the context for imports using `loadWith`.  Specifically, the `startingContext` field of the `Status` type should work for transitive imports.  If it doesn't then that sounds like a bug that should be fixed.

**Comment by hanshoglund** [comment]
I came across a related idea: a primitive that converts arbitrary expressions to the `Text`of their (normalized) Dhall representation. E.g. `toDhall : forall (a : Type) -> a -> Text`

@Gabriel439 do you think there would be any soundness/normalization issues with such a primitive?

Some context https://github.com/askvader/askvader/issues/1

**Comment by Gabriel439** [comment]
@hanshoglund: The main issue with that is that it would violate parametricity, meaning that there are fewer guarantees (free theorems) you get for polymorphic expressions.  The second issue is that it requires you to standardize how arbitrary expressions are rendered (i.e. should they be Unicode or ASCII, how much spacing to use, etc.).

**Comment by hanshoglund** [comment]
@Gabriel439 

Standard representation: That seems solvable, just requires picking a normal form of the syntax (e.g. the one already provided by the formatter).

Parametricity: Good point. So this is basically the [seq problem](https://www.google.com/search?q=Free+Theorems+in+the+Presence+of+seq). Of course, one could take the "put seq in a type class" approach, but that would preclude exporting anything involving functions (not a problem, in the JSON case, suggesting that the "create JSON strings in Dhall" is a sensible there - using dict passing in place type classes of course).

My use case for this was staging, e.g. consider top-level configuration expression for a computing cluster containing functions that we want to transfer to various nodes, where they are to be applied to runtime configuration values (example below). I guess this is not solvable without weakening parametricity.

```hs
let RuntimeInfo = { actualIp : Text, clusterSize : Natural, ... }
let ServerConf = ... 
let topLevel = { serverConfig : RuntimeInfo -> ServerConf, minSize : Natural, maxSize : Natural, ... }
```

**Comment by Gabriel439** [comment]
@hanshoglund: Could you expand a bit more on the staging use case?  The reason I ask is that it's not clear why you need to represent Dhall expressions as text literals Dhall in order to be able to stage them.

**Comment by hanshoglund** [comment]
@Gabriel439: My current implementation involves a custom evaluator (using the Haskell API) that evaluates part of the top-level expression (to obtain e.g. servers to boot), but leaves other parts unevaluated for later execution (e.g. to be copied to the booted servers as text). This works fine, but requires ad-hoc work on the Haskell side managing ASTs and so on. This is fair enough: I'm in a sense writing a custom (distributed) normalizer for Dhall expressions and the fact that I can do this by reusing most of the standard one through its Haskell API is great.

The idea behind `toDhall` was to represent the staging in Dhall itself, and thus removing the need for a custom evaluator. On the other hand the loss of parametricity indicate that this might not actually simplify things as much as I first thought. I'll probably stick to my original idea :).

**Comment by Gabriel439** [comment]
@akshaymankar: Would your original concern about using the `json_pure_dhall` branch be resolved if Dhall had a keyword for converting records to lists of key-value pairs?  See this comment for more details:

https://github.com/dhall-lang/dhall-lang/issues/234#issuecomment-422478261

If we had that, then you could convert a Dhall record to a JSON recordd, like this:

```haskell
let JSON =
      https://raw.githubusercontent.com/dhall-lang/dhall-lang/gabriel/json_pure_dhall/Prelude/JSON/package.dhall sha256:76a75f75e9ab6cd78d18b123f10946cc7778a3d734e50978d635505cae8c062a

let record =
      { foo =
          JSON.array [ JSON.bool True, JSON.number 1.0 ]
      , bar =
          JSON.string "!"
      }

let value = JSON.object (toMap record)

in  JSON.render value
```

**Comment by akshaymankar** [comment]
Would `toMap` be able to convert any record to assoc list or would it have to be a record with all the values of type `JSON`?
It would be great if it works on all record types, but I can already see a couple of implementation problems with it. For instance would it be an error if one of the values is a function? Implementing this kind of check could get tricky.
But if it doesn't work on _any_ record type, I am not sure if this solves the tediousness of converting a big record with recursive elements to JSON.

**Comment by Gabriel439** [comment]
@akshaymankar: I believe `toMap` should probably not automatically recurse because then there would be no way to opt out of the automatic recursion behavior for nested records.

However, I think the way I'm going to proceed is to try to convert a few JSON configuration formats to Dhall using the `json_pure_dhall` branch once the `Text/show` support is merged so I can get a better idea of what's necessary to make this easy.

**Comment by sjakobi** [comment]
Can this issue be closed now that we have both `toMap` and [the `JSON` utilities](https://github.com/dhall-lang/dhall-lang/tree/da02ab2faf81b3e627f01cc5f4bb37403a15ab0e/Prelude/JSON) in the Prelude?

**Comment by akshaymankar** [comment]
@sjakobi Yes, `toMap`, the JSON utilities and `dhall-to-json` solve all my problems. So, this can be closed :)

**Comment by anttih** [comment]
I'm writing dhall bindings for https://docs.drone.io and it has a concept of a plugin for a pipeline `step` definition. Depending on the plugin, the `settings` field may have a very different type. [The docker plugin](http://plugins.drone.io/drone-plugins/drone-docker/) for example might have `Text` or even nested maps as part of the settings map. I would like to have a way to gradually add type safety around certain plugins and let the "root" type just be `JSON` for the `settings` field.

I cannot create a union because I don't know all the plugins one might use. Nor can I use polymorphism for the `settings` field since one might be using multiple plugins in one pipeline. I don't see any other way to solve this than having a recursive `toMap`.

I see this pattern quite often. And once you add single `JSON` field to your configuration it often makes you write a whole lot of boilerplate for manually mapping values to `JSON`.

**Comment by Gabriel439** [comment]
@anttih: Just to make sure I understand the use case: you have some top-level type which might hypothetically look like this:

```dhall
let Pipeline = { steps : List Step, … }

let Step = { name : Text, image : Text, settings : ??? }
```

... and without any language changes the two main alternatives for how to type the `settings` field are:

* Use `Prelude.JSON.Type`

  ```dhall
  let Step = { name : Text, image : Text, settings : Prelude.JSON.Type }
  ```

  However, if you do that then you need to write a `…/toStep` function for each possible `Step` converting from a strongly-typed representation to `Prelude.JSON.Type` in order to verify that the strongly-typed representation makes invalid `Step`s unrepresentable

* Use polymorphism

  ```dhall
  let Step = λ(settings : Type) → { name : Text, image : Text, settings : settings }
  ```

  ... but then if you do that then you still need to convert the `settings` field to `Prelude.JSON.Type` anyway at the end if you want to stick them all inside the same `List`

So both approaches essentially boil down to the first approach: is there an ergonomic way to convert the `settings` field to to `Prelude.JSON.Type` without writing a custom conversion function for each possible `Step`?

Does that sound like what you are looking for?

**Comment by anttih** [comment]
@Gabriel439 Yes exactly.

**Comment by Gabriel439** [comment]
@anttih: Perhaps instead of a recursive `toMap` you might want a `toJSON` keyword that automatically converted a Dhall value to `Prelude.JSON.Type`?

**Comment by anttih** [comment]
@Gabriel439 yes, for some reason I thought they are quite the same but they are not. What I want is `toJson`.

**Comment by Gabriel439** [comment]
Alright, then I'll reopen an issue and propose changing the scope slightly from what was originally proposed.  The new proposal is to add a new `toJSON` keyword that converts to `Prelude.JSON.Type` instead of an abstract `JSON` type

**Comment by ari-becker** [comment]
This would be fantastic for `dhall-concourse`... we tried with unions and let's just say that [it started to turn into a mess](https://github.com/coralogix/dhall-concourse/blob/master/types/resources/_unions.dhall) (linking to it here to reference prior art / pain-points).

**Comment by Gabriel439** [comment]
This same functionality was also requested in a Stack Overflow Q&A: https://stackoverflow.com/q/61344067/1026598

**Comment by timbertson** [comment]
I'm in two minds about this. On the one hand, the pure dhall JSON type can be confusing with its recursive encoding (and can cause [performance problems](https://discourse.dhall-lang.org/t/figuring-out-performance-bottlenecks/251/10)). But making it opaque would have downsides, too:

 - loses the ability to control conversion
 - may make people design worse APIs

For the first point, I'm thinking about things like skipping record fields when their value is `null` / `[]` / `{}`. Currently I have implemented these myself for use in various `toJSON` functions, would they be exposed as options to the `toJSON` keyword?

The second point is maybe a philosophical one. Currently since I _need_ to have a `toJSON` function anyway, I can make nice canonical type design choices. As an example, I'm working on some github actions bindings. GHA are pretty loosely typed, e.g. each `step` can specify a `uses` field (action ID) and `with` (params). _Or_ it can have `run` with `env`. You can never pass `with` for a `run` step, nor can you pass `env` and `uses`. So that is much better modeled as a union, and even a nested union for the `uses` type (since that can be the name of a repo or a docker image ID)

If toJSON were added, there would be much more pressure to make my types match the destination JSON schema, i.e. flatten various things into `Text` instead of having nice structures types and unions. Because if I wanted to represent one leaf node differently (say as a proper `DockerImage` type instead of `Text`), I'd then have to create a whole stack of `toJSONStructure` functions mapping the typed version to a flat record version matching the desired output schema. _And_ make sure my users always used it.

I've noticed that all the open-source github actions take this approach, possibly because they're using `dhall to-json` instead of `dhall to-directory-tree`. So I guess just highlighting that making this convenient and opaque would encourage people to use weaker types.

This is probably getting way off track, but is there any possibility of associating a particular conversion with a type?

```
let Action = <
   , Docker: DockerImage
   , Action: Text
> with asJSON = \(action: Action) -> { ... }
```

I feel like I'm accidentally proposing typeclasses, which is probably a bit of an extreme reaction to this one awkwardness ;)

(..and that would probably also explode the size of the normalized form much like the current JSON type, now that I think about it :/)

**Comment by Gabriel439** [comment]
@timbertson: Note that the scope of the issue has changed so that the `JSON` type would no longer be opaque (i.e. it would remain the same as it currently is in the Prelude)

**Comment by timbertson** [comment]
Ah right, sorry. So for the first case, I'd have the ability to implement `omitEmptyValues` on top of the final JSON structure. It'd be possible but inconvenient to apply different options to individual parts of the JSON tree, but I haven't come across any need for that myself.

The issues around type design still stand I think, in that you're forced to write a stack of conversion functions if some leaf node doesn't match the output JSON schema.

**Comment by Gabriel439** [comment]
@timbertson: Not necessarily.  If there were an ergonomic way to update those leaf nodes using the `with` syntax, then you could do something like this:

```dhall
toJSON (exampleValue with deeply.nested.field  = f exampleValue.deeply.nested.field)
```

... then you would only need to define one conversion function.

Admittedly, the `with` function is currently impoverished (since it doesn't work through `Optional` or `List`, yet), but once it did then I think this might be viable for your use case.

**Comment by timbertson** [comment]
Ah true, yes I think that would work

**Comment by KAction** [comment]
FWIW, I would be glad if toJSON primitive function appear. As @Gabriel439 showed in answer to my [question](https://stackoverflow.com/questions/63641241/how-to-convert-structure-to-json-in-dhall), it is possible to convert arbitrary structure to json, but it is verbose and repetitive.

**Comment by alt-romes** [comment]
> Alright, then I'll reopen an issue and propose changing the scope slightly from what was originally proposed. The new proposal is to add a new `toJSON` keyword that converts to `Prelude.JSON.Type` instead of an abstract `JSON` type

Exactly! I would like to add my support towards this, it would be very useful for my current use case.

I can help with the implementation if given some pointers too.

Thanks!

**Comment by Gabriella439** [comment]
@alt-romes: Probably the best way to get started is to study a similar change that has been standardized before.  Probably the most similar change would be this PR that standardized `showConstructor`:

https://github.com/dhall-lang/dhall-lang/pull/1257

If you're still not sure where to begin, you can open a draft PR with only some sample test cases and then I can walk you through the rest of the process on the PR discussion

**Comment by michelk** [comment]
Is it now possible to easily convert dhall-records to json-strings within a dhall-document?

**Comment by bgohla** [comment]
> Is it now possible to easily convert dhall-records to json-strings within a dhall-document?

A generic one would be great. But the best you can do for now is still `Prelude.JSON.render` and `toMap` it seems.
